Screen Manager Microphone Input Manager (#944)
smartdevicelink/sdl_evolution: proposals/nnnn-screen-manager-audio-pass-thru.md

## Proposed solution
The proposed solution is to add a new private `SDLAudioPassThruManager` sub-manager to the screen manager. This manager will handle APT-related capabilities, text, callbacks, etc. The `SDLScreenManager` itself will then provide a simple public API for presenting an APT to the screen.

### Microphone Input View
##### iOS
```objc
@interface SDLMicrophoneInputView

/// Set this to change the default timeout for all microphone input. If a timeout is not set on an individual object (or if it is set to 0.0), then it will use this timeout instead. See `timeout` for more details. If this is not set by you, it will default to 10 seconds. The max is 1000 seconds.
@property (class, assign, nonatomic) NSTimeInterval defaultTimeout;

/// Maps to `PerformAudioPassThru.audioPassThruDisplayText1`. The primary text of the view.
@property (strong, nonatomic, nullable) NSString *text;

/// Maps to `PerformAudioPassThru.audioPassThruDisplayText2`. The secondary text of the view.
@property (strong, nonatomic, nullable) NSString *secondaryText;

> /// Maps to `PerformAudioPassThru.initialPrompt`. Will play a prompt to the user before microphone input is begun.

Reviewer: ```suggestion
/// Maps to `PerformAudioPassThru.initialPrompt`. Will play a prompt to the user before microphone input is started.
```

/// @param microphoneDataHandler A handler that will be called with raw audio data
/// @completionHandler A handler sent when the head unit sends a response and a possible error if one occurs.
- (void)presentMicrophoneInputView:(SDLMicrophoneInputView *)microphoneInputView withAudioDataHandler:(SDLMicrophoneDataHandler)microphoneDataHandler completionHandler:(nullable SDLScreenManagerUpdateHandler)completionHandler;

/// Dismisses the current microphone input view if there is one. If you're currently receiving audio data, you may not immediately cease to receive this data.
/// @param completionHandler A possible error if one occurs
- (void)dismissMicrophoneInputViewWithCompletionHandler:(nullable SDLScreenManagerUpdateHandler)completionHandler;

@end
```

##### Java
```java
public interface AudioDataListener {
    void onAudioData(byte[] audioData)
}

public class BaseScreenManager {
    public void presentMicrophoneInputView(@NonNull MicrophoneInputView view, @NonNull AudioDataListener audioDataListener, @Nullable CompletionListener completionListener)
    public void dismissMicrophoneInputView(@Nullable CompletionListener completionListener)
}
```

### JavaScript Suite APIs

> Due to the size of the iOS APIs and the similarity between the iOS, Java Suite and eventual JavaScript Suite APIs, this proposal does not present the public APIs of the JavaScript Suite APIs – especially because the JavaScript Suite APIs do not currently have a screen manager layer. The JavaScript Suite APIs should mirror the iOS and Java Suite API appropriately and is up to the maintainers' discretion. However, if any changes needed to be made such that they impacted the iOS / Java Suite API (such as the alteration, addition, or removal of a method or property), then a proposal revision would be needed.

Reviewer: ```suggestion
Due to the size of the iOS APIs and the similarity between the iOS, Java Suite and eventual JavaScript Suite APIs, this proposal does not present the public APIs of the JavaScript Suite APIs – especially because the JavaScript Suite APIs do not currently have a screen manager layer. The JavaScript Suite APIs should mirror the iOS and Java Suite API appropriately and is up to the Project Maintainer's discretion. However, if any changes needed to be made such that they impacted the iOS / Java Suite API (such as the alteration, addition, or removal of a method or property), then a proposal revision would be needed.
```
Please change "maintainers'" to be more specific, "Project Maintainer's"

### Additional Implementation Notes
- The internal alert manager will observe the `PerformAudioPassThru` response to know when the view has finished presenting, and then call the `completionHandler`.
- The internal manager will always send the RPC, even if the system context is not MAIN. If the `PerformAudioPassThru` returns a failure to present, it will call the `completionHandler` with the error.
- The developer will not be notified when the view appears on the screen, assuming no error occurred – see alternative #1 for possible ways to do that.
- The `SDLMicrophoneInputManager` sub-manager will use queues to manage related requests, similar to how the `SDLChoiceSetManager` does.
- If any files fail to upload, the presentation of the view should continue without an error.
- The view should be copied as soon as `presentMicrophoneInputView` is called in order to prevent the developer from changing the properties of the view after they call the method.

## Potential downsides

> The author can not think of any downsides to this proposal.

Reviewer: ```suggestion
The author cannot think of any downsides to this proposal.
```
